File: Arrays/RemoveDuplicatesFromSortedArray/code.java
/*
gfg : https://practice.geeksforgeeks.org/problems/remove-duplicate-elements-from-sorted-array/1


Given a sorted array A of size N, delete all the duplicates elements from A.


Example 1:

Input:
N = 5
Array = {2, 2, 2, 2, 2}
Output: 2
Explanation: After removing all the duplicates 
only one instance of 2 will remain.

Example 2:

Input:
N = 3
Array = {1, 2, 2}
Output: 1 2 

Your Task:  
You dont need to read input or print anything. Complete the function remove_duplicate() which takes the array A[] and its size N as input parameters and modifies it in place to delete all the duplicates. The function must return an integer X denoting the new modified size of the array. 
Note: The generated output will print all the elements of the modified array from index 0 to X-1.


Expected Time Complexity: O(N)
Expected Auxiliary Space: O(1)


Constraints:
1 <= N <= 10^4
1 <= A[i] <= 10^6
*/


class Solution {
    int remove_duplicate(int arr[],int n){
        // code here
        if(n==0 || n==1)
            return n;
        int j=0;
        for(int i=0;i<n-1;i++){
            if(arr[i] != arr[i+1])
                arr[j++] = arr[i];
        }
        arr[j++] = arr[n-1];
        return j;
   }
}

File: searchingAndSorting/quickSort/code.cpp
/*
gfg : https://practice.geeksforgeeks.org/problems/quick-sort/1/?track=dsa-workshop-1-search-sort&batchId=308#

Quick Sort is a Divide and Conquer algorithm. It picks an element as pivot and partitions the given array around the picked pivot.

Given an array arr[], its starting position low and its ending position high. Quick Sort is achieved using the following algorithm. 

quickSort(arr[], low, high)
{
    if (low < high)
    {
        /* pi is partitioning index, 
           arr[pi] is now at right place */
        pi = partition(arr, low, high);
        quickSort(arr, low, pi - 1);  // Before pi
        quickSort(arr, pi + 1, high); // After pi
    }
}
Implement the partition() function used in quickSort().


Example 1:

Input: 
N = 5 
arr[] = { 4, 1, 3, 9, 7}
Output: 1 3 4 7 9


Example 2:

Input:
N = 10
arr[] = {10, 9, 8, 7, 6, 5, 4, 3, 2, 1}
Output: 1 2 3 4 5 6 7 8 9 10

Your Task: 
You don't need to read input or print anything. Your task is to complete the function partition() which takes the array arr[], low and high as input parameters and partitions the array. Consider the last element as the pivot such that all the elements less than(or equal to) the pivot lie before it and the elements greater than it lie after the pivot. 


Expected Time Complexity: O(N)
Expected Auxiliary Space: O(1)


Constraints:
1 <= N <= 10^3
1 <= arr[i] <= 10^4
*/

void swap(int* a, int* b)  
{  
    int t = *a;  
    *a = *b;  
    *b = t;  
} 
int partition (int arr[], int low, int high)  
{  
    int pivot = arr[high]; // pivot  
    int i = (low - 1); // Index of smaller element  
  
    for (int j = low; j <= high - 1; j++)  
    {  
        // If current element is smaller than the pivot  
        if (arr[j] < pivot)  
        {  
            i++; // increment index of smaller element  
            swap(&arr[i], &arr[j]);  
        }  
    }  
    swap(&arr[i + 1], &arr[high]);  
    return (i + 1);  
} 
